src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala

```diff
@@ -0,0 +1,511 @@
+          )
+          if (relevantLines.nonEmpty) {
+            println("[SUBMIT] Relevant error lines:")
+            relevantLines.foreach(line => println(s"[SUBMIT]   $line"))
+          }
+
+          throw new RuntimeException(s"Spark submit failed with exit code ${result.exitCode}")
+        }
+      } else {
+        if (attempt > 1) {
+          println(s"[SUBMIT] Job submitted successfully on attempt $attempt")
+        } else {
+          println(s"[SUBMIT] Job submitted successfully")
+        }
+        result
+      }
+    }
+
+    attemptSubmit()
+  }
+
+  private def watchJob(
+      queueName: String,
+      jobSetId: String,
+      config: TestConfig,
+      context: TestContext
+  ): Future[TestResult] = Future {
+    val podMonitor = new SimplePodMonitor(context.namespace)
+
+    println(s"[WATCH] Starting job watch for jobSetId: $jobSetId")
+    val jobFuture = armadaClient.watchJobSet(queueName, jobSetId, jobWatchTimeout)
+
+    var jobCompleted               = false
+    var podFailure: Option[String] = None
+    var assertionResults           = Map.empty[String, AssertionResult]
+
+    if (config.failFastOnPodFailure) {
+      println(s"[MONITOR] Starting pod monitoring for namespace: ${context.namespace}")
+      val monitorThread = new Thread(() => {
+        while (!jobCompleted && podFailure.isEmpty) {
+          podFailure = podMonitor.checkForFailures()
+          if (podFailure.isEmpty) {
+            Thread.sleep(5000) // Check every 5 seconds
+          }
+        }
+      })
+      monitorThread.setDaemon(true)
+      monitorThread.start()
+    }
+
+    // If we have assertions, run them after driver starts but while job is running
+    val assertionThread = if (config.assertions.nonEmpty) {
+      val thread = new Thread(() => {
+        // Wait for driver to start, then run assertions while job is active
+        assertionResults = runAssertionsWhileJobRunning(
+          config.assertions,
+          context,
+          jobCompleted = () => jobCompleted
+        )
+      })
+      thread.setDaemon(true)
+      thread.start()
+      Some(thread)
+    } else {
+      None
+    }
+
+    // Wait for job to complete or fail
+    val jobStatus =
+      try {
+        import scala.concurrent.Await
+        Await.result(jobFuture, jobWatchTimeout + 10.seconds)
+      } catch {
+        case _: TimeoutException =>
+          println(s"[TIMEOUT] Job watch timed out after ${jobWatchTimeout.toSeconds}s")
+          JobSetStatus.Timeout
+        case ex: Exception =>
+          println(s"[ERROR] Job watch failed: ${ex.getMessage}")
+          JobSetStatus.Failed
+      } finally {
+        jobCompleted = true
+      }
+
+    // Wait for assertion thread to complete if it exists
+    assertionThread.foreach { thread =>
+      try {
+        thread.join(30000) // Wait up to 30 seconds for assertions to complete
+      } catch {
+        case _: InterruptedException =>
```

Comment:
I'm not sure I understand this code.  If the assertion thread gets interupted, how will you know if there were any failures?